src/main/java/de/tum/in/test/api/util/ClassMemberAccessor.java

```diff
@@ -0,0 +1,213 @@
+	 *                       private or protected (inherited) methods.
+	 * @param parameterTypes The parameter types of this method.
+	 * @return The wanted method.
+	 * @throws NoSuchMethodException Thrown if the specified method cannot be found.
+	 */
+	static Method getMethod(Class<?> declaringClass, String methodName, boolean findPrivate, Class<?>[] parameterTypes)
+			throws NoSuchMethodException {
+		if (!findPrivate) {
+			try {
+				return declaringClass.getMethod(methodName, parameterTypes);
+			} catch (NoSuchMethodException nsme) {
+				// Also search for declared methods in the own class even when not explicitly
+				// searching for private methods to be able to provide error messages to the
+				// users that the method exists, but with the wrong visibility.
+				return declaringClass.getDeclaredMethod(methodName, parameterTypes);
+			}
+		} else {
+			return getPrivateMethod(declaringClass, methodName, parameterTypes);
+		}
+	}
+
+	/**
+	 * Retrieve a method with arguments of a given class by its name.
+	 * <p>
+	 * Also recursively searches for an inherited method in all superclasses and
+	 * implemented interfaces. Finds a method even if it is not declared to be
+	 * visible outside the declaring class.
+	 *
+	 * @param declaringClass The class that declares or inherits the method.
+	 * @param methodName     The name of the method.
+	 * @param parameterTypes The parameter types of this method.
+	 * @return The wanted method.
+	 * @throws NoSuchMethodException Thrown if the specified method cannot be found.
+	 */
+	private static Method getPrivateMethod(Class<?> declaringClass, String methodName, Class<?>[] parameterTypes)
+			throws NoSuchMethodException {
+		return getClassHierarchy(declaringClass).flatMap(c -> {
+			try {
+				return getInheritedMethod(declaringClass, c, methodName, parameterTypes);
+			} catch (NoSuchMethodException nsme) {
+				return Stream.empty();
+			}
+		}).findFirst().orElseThrow(() -> new NoSuchMethodException(methodName));
+	}
+
+	/**
+	 * Searches for a method in the target class that might be inherited from the
+	 * declaring class.
+	 *
+	 * @param targetClass    The class in which the method should be accessible.
+	 * @param declaringClass The class from which the method might be inherited.
+	 * @param methodName     The name of the method.
+	 * @param parameterTypes The parameter types of the method.
+	 * @return A method that is accessible in the target class.
+	 * @throws NoSuchMethodException Thrown if no method as specified could be found
+	 *                               in either class.
+	 */
+	private static Stream<Method> getInheritedMethod(Class<?> targetClass, Class<?> declaringClass, String methodName,
+			Class<?>[] parameterTypes) throws NoSuchMethodException {
+		Method method = declaringClass.getDeclaredMethod(methodName, parameterTypes);
+		if (isInheritable(targetClass, declaringClass, method.getModifiers(), true)) {
+			return Stream.of(method);
```

Comment:
Would it be cleaner to just return the method or null and filter all null values in getPrivateMethod?

Reply:
I thought about that, too. But more compared to Optional. However, since this is all internal and used to flatten the stream, I thought it would be acceptable.

Reply:
But having both a checked exception any optional/stream/null might be suboptimal.

Reply:
The corresponding method for fields already used `Optional`s which is the better type for this purpose. I changed the Stream to Optional here as well.